### PR TITLE
chore (pkg/scale): Integrate scale into grandpa library

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

- Integrate new scale pkg into grandpa

## Tests

```
go test ./lib/grandpa -short -v
```

## Issues

- #1621

### CODE REVIEW

**Comment by jimjbrettj** [comment]
Waiting to open this until scale network PR is merged

File: lib/grandpa/network.go [hunk 1]
```diff
@@ -225,41 +225,20 @@ func (s *Service) sendNeighbourMessage() {
...
-	)
-
-	switch msg.Data[0] {
-	case voteType:
-		r := &bytes.Buffer{}
-		_, _ = r.Write(msg.Data[1:])
-		vm := &VoteMessage{}
-		err = vm.Decode(r)
-		m = vm
+	err = scale.Unmarshal(msg.Data, &gossipMessage)
```
Comment (EclesioMeloJunior) [inline lib/grandpa/network.go:228]: we need to check this error, right?

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: lib/grandpa/network.go (hunk 1 above)
Comment (jimjbrettj) [reply lib/grandpa/network.go:228]: I have it below the switch, but I can move it up if that's better

Comment (EclesioMeloJunior) [reply lib/grandpa/network.go:228]: would be nice to check the error before the switch, will be easier to relate the error to the `Unmarshal`

File: lib/grandpa/grandpa.go [hunk 2]
```diff
@@ -842,7 +843,7 @@ func (s *Service) finalise() error {
 		return err
 	}
 
-	pcj, err := newJustification(s.state.round, bfc.Hash, bfc.Number, pcs).Encode()
+	pcj, err := scale.Marshal(newJustification(s.state.round, bfc.Hash, bfc.Number, pcs))
```
Comment (timwu20) [inline lib/grandpa/grandpa.go:846]: nit: create a local var for the justification on another line

File: lib/grandpa/message.go [hunk 3]
```diff
@@ -105,31 +72,17 @@ type VoteMessage struct {
...
-	return err
-}
-
 // Type returns voteType
 func (v *VoteMessage) Type() byte {
 	return voteType
 }
 
+// Index Returns VDT index
+func (v *VoteMessage) Index() uint { return 0 }
```
Comment (timwu20) [inline lib/grandpa/message.go:81]:
```suggestion
func (vm *VoteMessage) Index() uint { return 0 }
```
Update all the other receiver functions too.

File: lib/grandpa/message.go [hunk 4]
```diff
@@ -147,9 +100,12 @@ type NeighbourMessage struct {
 	Number  uint32
 }
 
+// Index Returns VDT index
+func (m *NeighbourMessage) Index() uint { return 2 }
```
Comment (timwu20) [inline lib/grandpa/message.go:104]:
```suggestion
func (nm *NeighbourMessage) Index() uint { return 2 }
```

File: lib/grandpa/message.go [hunk 5]
```diff
@@ -202,57 +135,8 @@ type CommitMessage struct {
...
-		err = f.AuthData[i].Decode(r)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
+// Index Returns VDT index
+func (f *CommitMessage) Index() uint { return 1 }
```
Comment (timwu20) [inline lib/grandpa/message.go:139]:
```suggestion
func (cm *CommitMessage) Index() uint { return 1 }
```

File: lib/grandpa/network.go (hunk 1 above)
Comment (timwu20) [reply lib/grandpa/network.go:228]: Yes please handle the error immediately

File: lib/grandpa/network.go [hunk 6]
```diff
@@ -33,6 +31,7 @@ var (
 	grandpaID                protocol.ID = "/paritytech/grandpa/1"
 	messageID                            = network.ConsensusMsgType
 	neighbourMessageInterval             = time.Minute * 5
+	gossipMessage                        = scale.MustNewVaryingDataType(&VoteMessage{}, &CommitMessage{}, &NeighbourMessage{}, &catchUpRequest{}, &catchUpResponse{})
```
Comment (timwu20) [inline lib/grandpa/network.go:34]: don't use the same VaryingDataType in `decodeMessage`.  This isn't thread safe if there are parallel calls to `decodeMessage`. You will be overwriting the same address repeatedly, which could lead to undesired side effects.

**Review by timwu20 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**